File: 2-add-two-numbers/add-two-numbers.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        vector<int> a1, a2;
        for (ListNode* i = l1; i; i = i->next) a1.push_back(i->val);
        for (ListNode* i = l2; i; i = i->next) a2.push_back(i->val);

        int i = 0;
        int j = 0;

        int res = 0, remain = 0;
        vector<int> ans;

        while (i < a1.size() && j < a2.size()) {
            int tmp = a1[i++] + a2[j++] + remain;
            res = tmp % 10;
            remain = tmp / 10;
            ans.push_back(res);
        }

        while (i < a1.size()) {
            int tmp = a1[i++] + remain;
            res = tmp % 10;
            remain = tmp / 10;
            ans.push_back(res);
        }
        while (j < a2.size()) {
            int tmp = a2[j++] + remain;
            res = tmp % 10;
            remain = tmp / 10;
            ans.push_back(res);
        }
        if (remain) ans.push_back(1);

        ListNode* head = nullptr;
        for (int i = ans.size() - 1; i >= 0; i--) {
            ListNode* tmp = new ListNode{ans[i], nullptr};
            if (!head) head = tmp;
            else {
                tmp->next = head;
                head = tmp;
            }
        }

        return head;
    }
};

File: 2-add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        ListNode cur1 = l1, cur2 = l2;
        ListNode head = null;
        ListNode tail = null;
        int more = 0;

        while (cur1 != null && cur2 != null) {
            int sum = cur1.val + cur2.val + more;
            more = sum / 10;
            int val = sum % 10;
            
            if (head == null && tail == null) {
                head = tail = new ListNode(val);
            } else {
                ListNode newNode = new ListNode(val);
                tail.next = newNode;
                tail = newNode;
            }

            cur1 = cur1.next;
            cur2 = cur2.next;
        }

        while (cur1 != null) {
            int sum = cur1.val + more;
            more = sum / 10;
            int val = sum % 10;

            ListNode newNode = new ListNode(val);
            tail.next = newNode;
            tail = newNode;

            cur1 = cur1.next;
        }

        while (cur2 != null) {
            int sum = cur2.val + more;
            more = sum / 10;
            int val = sum % 10;

            ListNode newNode = new ListNode(val);
            tail.next = newNode;
            tail = newNode;

            cur2 = cur2.next;
        }

        if (more > 0) {
            ListNode newNode = new ListNode(more);
            tail.next = newNode;
            tail = newNode;
        }

        return head;
    }
}